### PR TITLE
Solution for CF-963D

### PR DESCRIPTION
**Pull request by dutinmeow**

### CODE REVIEW

**Review by nathangong — APPROVED** [review]

**Comment by dutinmeow** [comment]
> you know what a constructor is
why not actually
use it

Using an init/build function is pretty normal in CP ([example](https://github.com/bqi343/USACO/blob/master/Implementations/content/data-structures/1D%20Range%20Queries%20(9.2)/SegTree%20(9.2).h)). Anyhow, I don't feel like it makes the code any more confusing or unclear

**Comment by SansPapyrus683** [comment]
... ok but why don't you want to put it in a constructor?

File: solutions/cf-963D.mdx [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		int next[26];
+		State(int len = 0, int link = -1, int pos = -1) : len(len), link(link), pos(pos) {
+			memset(next, -1, sizeof(next));
+		}
+	};
+ 
+	vector<State> states;
+	
+	SuffixAuto() {}
+
```
Comment (SansPapyrus683) [inline solutions/cf-963D.mdx:50]:
```suggestion
```
why does this still exist lol

File: solutions/cf-963D.mdx [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	int encode(char c) { return c - 'a'; }
+ 
+	inline int state(int len = 0, int link = -1, int pos = -1) {
+		states.emplace_back(len, link, pos);
+		return states.size() - 1;
+	}
+};
+//EndCodeSnip
+
+const int MAXN = 1e5 + 1;
```
Comment (SansPapyrus683) [inline solutions/cf-963D.mdx:108]: does it run w/o the +1?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

**Comment by dutinmeow** [comment]
> does it run w/o the +1?
You yourself said it was one-based

**Comment by dutinmeow** [comment]
> why does this still exist lol

A default constructor is necessary because I declare the Suffix Automaton in the global scope before the string is inputted. It would be a syntax error if it wasn't there (or I could put the construct method in an `init` function...)

**Comment by SansPapyrus683** [comment]
i'll just believe you
will review later, ping me if i forget

**Comment by dutinmeow** [comment]
how much later

File: solutions/cf-963D.mdx (hunk 1 above)
Comment (Aadit-Ambadkar) [reply solutions/cf-963D.mdx:50]: you need a constructor...

File: solutions/cf-963D.mdx [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	
+string S;
+SuffixAuto sa;
+ 
+int Q, ans[MAXN];  // answer for query
+vector<pii> queries[MAXN];  // queries per state
+vector<int> suffix_link_tree[MAXN * 2];  // suffix link tree (at most 2 * |S| nodes)
```
Comment (Aadit-Ambadkar) [inline solutions/cf-963D.mdx:115]:
```suggestion
// answer for query
int Q, ans[MAXN];  
// queries per state
vector<pii> queries[MAXN];  
// suffix link tree (at most 2 * |S| nodes)
vector<int> suffix_link_tree[MAXN * 2]; 
```

The last line was over 80 characters.

File: solutions/cf-963D.mdx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+ 
+vector<int> ord;  // visited nodes in dfs
+ 
+void dfs(int u = 0) {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-963D.mdx:119]:
```suggestion
void dfs(int u) {
```

File: solutions/cf-963D.mdx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		int s = sa.run(P);
+		if (s == -1) {
+			ans[i] = -1;
+		} else {
+			queries[s].EB(i, K);
+			ans[i] = P.size();
+		}
+	}
+ 
+	dfs();
```
Comment (Aadit-Ambadkar) [inline solutions/cf-963D.mdx:165]:
```suggestion
	dfs(0);
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/cf-963D.mdx (hunk 1 above)
Comment (Aadit-Ambadkar) [reply solutions/cf-963D.mdx:50]: You still need a constructor

Comment (SansPapyrus683) [reply solutions/cf-963D.mdx:50]: no he declared one @ the bottom
dustin explained in one of his comments

**Review by Aadit-Ambadkar — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
i'll wait for someone who actually knows about suffixes to review before i merge

**Pull request merged**